[WGSL] Always visit vertex entry point output

tadeuzagallo opened the pull request:
#### 8298a862fed61d92b48e246066f4968c43fe8bcb
<pre>
[WGSL] Always visit vertex entry point output
<a href="https://bugs.webkit.org/show_bug.cgi?id=251268">https://bugs.webkit.org/show_bug.cgi?id=251268</a>
&lt;rdar://problem/104744731&gt;

Reviewed by Myles C. Maxfield.

Right now, in order to translate the struct member attributes correctly to metal,
we check whether a struct is being used as an output of the vertex entry point
(in order to use the `[[user()]]` attributes to connect it with the fragment entry
point. This is a bit hacky, but works for now. However, the bug was that we didn&apos;t
perform this check if we didn&apos;t need to rewrite the entry point&apos;s input arguments,
but the two things are unrelated. This had broken the hello-triangle demo.

Also fixed the FIXME in the example since we now support binary add.

* Source/WebGPU/WGSL/EntryPointRewriter.cpp:
(WGSL::EntryPointRewriter::rewrite):
(WGSL::EntryPointRewriter::checkReturnType):
* Websites/webkit.org/demos/webgpu/scripts/hello-triangle.js:
(async helloTriangle):

Canonical link: <a href="https://commits.webkit.org/259546@main">https://commits.webkit.org/259546@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/95d46606bedb7f193e8bef6086028eef83b8cdf3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105031 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14110 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37921 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114294 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15247 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5032 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97350 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113307 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110787 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11785 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94785 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39296 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93654 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26412 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80960 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7448 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27771 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7542 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4353 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13597 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47323 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6573 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9331 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WGSL/EntryPointRewriter.cpp:
```diff
@@ -59,6 +59,7 @@ class EntryPointRewriter {
     static AST::TypeDecl& getResolvedType(AST::TypeDecl&);
```
from the commit message:

> we check whether a struct is being used as an output of the vertex entry point

Can we just always use a struct for inputs and outputs unconditionally, even if they're empty / only have one member? There's no reason to _not_ just always use them. It would make the compiler's internal logic simpler, at the cost of slightly more characters in the generated output. I think that tradeoff is worth it.

tadeuzagallo replied:
Converting non-struct outputs to struct also isn't free (implementation wise)... it requires modifying the AST to 1) declare the struct 2) populate the struct. Maybe I'm just looking at it from the opposite perspective, but I feel like that's more complicated than checking the type of the output, with the side effect of emitting more code. I don't feel very strong about it though, and I'm still understanding the constraints of whether those structs can be used in multiple contexts, etc. so I'm not opposed to revisiting it later.

tadeuzagallo committed:
[WGSL] Always visit vertex entry point output
https://bugs.webkit.org/show_bug.cgi?id=251268
<rdar://problem/104744731>

Reviewed by Myles C. Maxfield.

Right now, in order to translate the struct member attributes correctly to metal,
we check whether a struct is being used as an output of the vertex entry point
(in order to use the `[[user()]]` attributes to connect it with the fragment entry
point. This is a bit hacky, but works for now. However, the bug was that we didn't
perform this check if we didn't need to rewrite the entry point's input arguments,
but the two things are unrelated. This had broken the hello-triangle demo.

Also fixed the FIXME in the example since we now support binary add.

* Source/WebGPU/WGSL/EntryPointRewriter.cpp:
(WGSL::EntryPointRewriter::rewrite):
(WGSL::EntryPointRewriter::checkReturnType):
* Websites/webkit.org/demos/webgpu/scripts/hello-triangle.js:
(async helloTriangle):

Canonical link: https://commits.webkit.org/259546@main
